### PR TITLE
[ci skip] System-test: Begin list of needed high-level tests

### PR DESCRIPTION
**Pull request by cevich**
Signed-off-by: Chris Evich <cevich@redhat.com>

### CODE REVIEW

**Comment by cevich** [comment]
@edsantiago @ypu PTAL what I have so far.  This can be added to more later, it is just a starting place.  I tried to write workflows that will self-validate (i.e. final test commands prove former commands were successful).

**Comment by ypu** [comment]
Hi @cevich 
Checked the items in the doc. Seems we can just copy some tests code from e2e directly except the last one (help). The test case under e2e is named with the subcommand. We need update the AfterEach() and BeforeEach() and adjust the steps(may remove some prepare steps as the test cases in e2e is separated) to fit the test cases.
For example, the second test case for stop and wait can be build with following test cases in e2e:
  - podman create container based on a remote image
  - podman start single container by name
  - podman exec simple command
  - podman logs for container
  - Podman wait(need some update for status and wait time)

Need add podman bud related functions like copy Dockerfile and update the Dockerfile and so on. But we already have these functions buildah testsuit. So it should not be hard.

I will take 7 days vacation in China for our spring festival from tomorrow. So I will not have time to init the pullreq until then. Maybe we can start with test case two as all needed common functions are ready. And I will start work on the bud related functions and other test cases as soon as I come back. Thanks for understanding.

**Comment by cevich** [comment]
@ypu thanks for taking a look, and for letting us know about the spring festival.

I chatted with @edsantiago some more on this.  We both still think golang and ginkgo (in general) are not a good fit for what the system-tests need to do.  They're going to gate at the RPM/packaging level, so there is very little, and very slow, feedback when they break.  So simplicity is key, I even wonder if maybe BATS is too much.  Maybe we can just have a simple shell-script.

For this PR, it is only about making a list of what we need.  I will hopefully have that finished before you get back.  Ed said he will work on adding a few starting/basic system-tests in another PR.  So we can get the packaging side of this up and working.  Then we can work to add more system-test later from this 'TODO' list.

**Comment by ypu** [comment]
Hi @cevich 
I feel I start to understand why you keep saying that golang and ginkgo is too high level. Seems you want to find a way that can test the binary directly in a system without any further request such like python dependence and bats support and so on. So the user can test podman like they using it in their env quickly.

But in my opinion this is why we should using ginkgo. 
First of all, system test is not the unit test. So it should be run under a certain system which means the rpm or package of podman should be build in a build system already. And the tests is not request more package than the podman binary except ginkgo which can be get easily with go get. So it has very low chance this will break the build process.

Second, if you look deep into BDD style testing, you should see it try to separate the test scenario and test code. So everyone can add test cases to it even for who can not coding. Ginkgo may not doing that in a very obvious way like other BDD style test framework which has separated feature file and code. But we have our e2e test for it. Each e2e test is a simple test step and if you grep the e2e test "It" items you can get all operates we already support, and people can make up system tests with simple copy and paste with it. Or users can just provide some description of the test case like what you did in this doc and we can translate it into a case quickly based what we already have in e2e tests.

So this is my reason for choice ginkgo for system test. As seems your pullreq for bash is still not ready. Maybe I can have a try for this? Looking forward for your feedback. :)

**Comment by cevich** [comment]
@ypu yes your understanding is correct.  If we are testing user running shell commands, then test should just run shell commands.  There is no better way to run shell commands than...using a shell :smile: 

I know about BDD.  The middle term, "Driven" is key:  You're suppose to write the tests BEFORE writing the code, so we are using it backwards without this benefit.  I also found the syntax-sugar of ginko/gomega to be very confusing and get in the way.  It does not follow standard Gherkin terms or feature descriptions.  Worse, the deeply nested-function style breaks up the scenario rules such that it's very hard for a newcomer to read/understand.

I agree with you debugging tests requires understanding them to some degree.  However, with so many tests and different people working on them, the extraneous ginkgo (and BDD in general) syntax seems to just get in the way.  It may be fine to use at the integration-test level, I'm not arguing that.

Rather, we want system-test to be as simple and boring as possible.  We should not spend lots of time/effort on them, nor should any future (new) developer.  We must focus most of our efforts writing/testing at the unit and integration level.  System-test is simply final gate before release.  It should always pass.  If system-tests ever fail, they should fail because there is serious/major problem we did not catch at unit/integration level.

Simply put:  System-test failure -> write new/better unit or integration test

Still, we are not 100% committed to BATS or similar framework yet, this PR is only about writing notes about requirements and planning :smile:

**Comment by cevich** [comment]
lolz...ci/prow/validate is suppose to fail...my commits are messed up here, fixing...

**Comment by cevich** [comment]
@TomSweeneyRedHat @umohnani8 Mind taking a quick peek, to see if there are any glaring typos here?

This is just high-level documentation and a TODO list, no actual code, only my ramblings :smile:
(Ed's PR #2533 backs this PR up with some actual code)

File: test/system/TODO.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
+![PODMAN logo](../../logo/podman-logo-source.svg)
+
+# Overview
+
```
Comment (TomSweeneyRedHat) [inline test/system/TODO.md]: 'Podman' and not 'podman' except for actual commands.

File: test/system/TODO.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+The system-tests must execute cleanly on all tested platforms.  They may optionally
+be executed during continuous-integration testing of code-changes, after all other
+testing completes successfully.  For a list of tested platforms, please see [the
+CI configuration file.](../../.cirrus.yml)
+
+
+# Execution
+
+When working from a clone of [the libpod repository](https://github.com/containers/libpod),
+the main entry-point for humans and automation, is `make localsystem`.  When operating
```
Comment (TomSweeneyRedHat) [inline test/system/TODO.md]: I'd remove the comma after automation

Comment (TomSweeneyRedHat) [reply test/system/TODO.md]: and actually the other in this sentence, be brave, go comma free!

File: test/system/TODO.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+from a packaged version of the system-tests, the entry-point may vary and therefore
+must be documented for the platform.  Running the packaged system-tests assumes
+a matching version of podman, and all standard dependencies are installed.
+
+
+# Test Design and overview
+
+System-tests should be high-level and user work-flow oriented.  For example, consider
+how multiple podman invocations would be used together by an end-user.  This collection
+of related commands, and the end-result would be considered a single test.  If one or
```
Comment (TomSweeneyRedHat) [inline test/system/TODO.md]: I'm on a comma killing mission apparently, I'd go with "related commands and"

File: test/system/TODO.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  - Import tarball
+  - Tag imported image
+  - Run imported image to confirm tarball modification, block on non-special signal
+  - Kill can send non-TERM/KILL signal to container to exit
+  - Confirm exit within timeout
+
+- [ ] Container runlabel, exists, checkpoint, exists, restore, stop, prune
+
+  - Using pre-existing remote image, start it with 'podman container runlabel --pull'
+  - Run a named container that exits immediatly
```
Comment (TomSweeneyRedHat) [inline test/system/TODO.md]: s/immeditly/immediately/

File: test/system/TODO.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+- save
+- search
+- stats
+- top
+- umount, unmount
+- unpause
+- volume
+- `--namespace`
+- `--storage-driver`
+
```
Comment (TomSweeneyRedHat) [inline test/system/TODO.md]: You obviously have jira cards in place for all of the above...

File: test/system/TODO.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+***Note***: Common operations for intra-test cleanup/reset have been omitted as
+are verified simply by executing the systemtests.
```
Comment (TomSweeneyRedHat) [inline test/system/TODO.md]: s/systemtests/system tests/

Comment (TomSweeneyRedHat) [inline test/system/TODO.md]: s/as are/as they are/

File: test/system/TODO.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+how multiple podman invocations would be used together by an end-user.  This collection
+of related commands, and the end-result would be considered a single test.  If one or
+more of the intermediate commands fail, the test could still pass as long as the
+end-result is still achieved.
+
+# Matrix:
+
+- As-installed defaults, root user
+- As-installed defaults, regular user
+
```
Comment (TomSweeneyRedHat) [inline test/system/TODO.md]: Maybe it will be clearer once you get the TODO's done, but I'm not sure what these two lines in the Matrix section stand for.

File: test/system/TODO.md (hunk 6 above)
Comment (TomSweeneyRedHat) [inline test/system/TODO.md]: You're probably aware, but just in case.  Buildah has a decent Docker 'conformance" [test suite](https://github.com/containers/buildah/tree/master/tests/conformance).   IDK if you can or would want to leverage that.

File: test/system/TODO.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  - Confirm 'container exists' zero exit (both containers)
+  - Checkpoint the running container
+  - Confirm 'container exists' non-zero exit (runlabel container)
+  - Confirm 'container exists' zero exit (named container)
+  - Run 'container restore'
+  - Confirm 'container exists' zero exit (both containers)
+  - Stop container
+  - Run 'container prune'
+  - Confirm podman ps -a lists no containers
+
```
Comment (TomSweeneyRedHat) [inline test/system/TODO.md]: maybe that will come in with the login/logout tests, but I didn't see much authentication kind of tests in the above.

**Comment by TomSweeneyRedHat** [comment]
Good start @cevich As I'm sure you're aware, I think you just scratching the surface with the number and types of tests we should have.  Just a few suggested grammar things through out.

**Comment by cevich** [comment]
@TomSweeneyRedHat Thank you so much for taking a look, yes just a starting-place.  And thanks for the reminder about adding cards, I'll add doing that, to my higher-priority TODO list :smile:

File: test/system/TODO.md (hunk 1 above)
Comment (cevich) [reply test/system/TODO.md]: k, got 'em.

File: test/system/TODO.md (hunk 2 above)
Comment (cevich) [reply test/system/TODO.md]: lol, but I love my comma's.  They're sooooo soft!

File: test/system/TODO.md (hunk 3 above)
Comment (cevich) [reply test/system/TODO.md]: k, got it.

File: test/system/TODO.md (hunk 7 above)
Comment (cevich) [reply test/system/TODO.md]: Oh, we don't need this.  I'll just remove the whole section.

File: test/system/TODO.md (hunk 6 above)
Comment (cevich) [reply test/system/TODO.md]: The system-tests will happen way downstream.  So we can simply assume those tests already passed.  That note is more about describing why we don't have system-tests dedicated to verifying cleanup.  I clarified it to be...more clear :smile:

File: test/system/TODO.md (hunk 8 above)
Comment (cevich) [reply test/system/TODO.md]: Those tests are really hard to do because they're either complex, or they rely heavily on external resources.  We need them, but best cover the basics first.   Both `login` and `logout` are in the TODO list.

**Comment by cevich** [comment]
@TomSweeneyRedHat okay, got all your suggestions and even opened a jira card:  RUN-303 :smile:   PTAL when you get a chance.  Thanks.

**Comment by mheon** [comment]
/approve
LGTM
@TomSweeneyRedHat mind giving one more ack so we can merge?
